lib/tbot/services/application/proxy_service.go

```diff
@@ -0,0 +1,305 @@
+	if err != nil {
+		return nil, nil, trace.Wrap(err)
+	}
+
+	s.log.DebugContext(ctx, "Requesting issuance of certificate for ProxyService proxy.")
+	routedIdent, err := s.identityGenerator.Generate(ctx, append(identityOpts, identity.WithRouteToApp(route))...)
+	if err != nil {
+		return nil, nil, trace.Wrap(err)
+	}
+	s.log.InfoContext(ctx, "Certificate issued for ProxyService proxy.")
+
+	return routedIdent.TLSCert, app, nil
+}
+
+func (s *ProxyService) startProxy(ctx context.Context) error {
+	// Ping the Teleport Proxy
+	proxyPing, err := s.proxyPinger.Ping(ctx)
+	if err != nil {
+		return err
+	}
+
+	// Retrieve the Proxy Address to use for the Application Request
+	proxyAddr, err := proxyPing.ProxyWebAddr()
+	if err != nil {
+		return err
+	}
+	s.proxyAddr = proxyAddr
+
+	// Build the proxy url to use in the http Client.
+	proxyUrl, err := url.Parse("https://" + proxyAddr)
+	if err != nil {
+		return err
+	}
+	s.proxyUrl = proxyUrl
+
+	// This router expects the requests to come in via the style "GET <fqdn>:<port>/"
+	// It doesn't really consider the CONNECT method, but it should work nonetheless.
+	proxyHttpServer := http.Server{
+		Addr:              s.cfg.Listen,
+		ReadTimeout:       apidefaults.DefaultIOTimeout,
+		ReadHeaderTimeout: defaults.ReadHeadersTimeout,
+		WriteTimeout:      apidefaults.DefaultIOTimeout,
+		IdleTimeout:       apidefaults.DefaultIdleTimeout,
+	}
+	proxyHttpServer.Handler = http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
+		err := s.handleProxyRequest(w, req)
+		if err != nil {
+			s.handleProxyError(w, err)
+			ctx.Done()
+			return
+		}
+	})
+
+	err = proxyHttpServer.ListenAndServe()
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (s *ProxyService) handleProxyError(w http.ResponseWriter, err error) {
+	trace.WriteError(w, err)
+	s.log.Error("Encountered an error while proxying request", "error", err)
+}
+
+func (s *ProxyService) handleProxyRequest(w http.ResponseWriter, req *http.Request) error {
+	// Resolve Application Name via either URL or Host Header
+	appName := cmp.Or(req.URL.Host, req.Header.Get("Host"))
+
+	ctx := req.Context()
+
+	var appCert *tls.Certificate
+	var err error
+	var cached bool
+
+	if s.cfg.CertificateCaching {
+		cached = true
+		appCert, err = utils.FnCacheGet(ctx, s.cache, appName, func(ctx context.Context) (*tls.Certificate, error) {
+			cached = false
+			s.log.InfoContext(ctx, fmt.Sprintf("(Re)issuing application Certificate for %s\n", appName))
+			cert, _, err := s.issueCert(ctx, appName)
+			return cert, err
+		})
+	} else {
+		appCert, _, err = s.issueCert(ctx, appName)
+	}
+
+	if err != nil {
+		return err
+	}
+
+	httpClient := &http.Client{
+		Transport: &http.Transport{
+			TLSClientConfig: &tls.Config{
+				Certificates:       []tls.Certificate{*appCert},
+				InsecureSkipVerify: s.botClient.Config().InsecureSkipVerify,
+			},
+		},
+	}
+
+	// Build the Application Request
+	upstreamRequest := http.Request{
```

Comment:
We should also ensure we propagate the context from the incoming request to the outgoing request so that if the client goes away, we aren't sat around waiting for the server to respond.